[PyTorch Edge][type] Add type support for NamedTuple custom class (import)

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62612
* __->__ #63130



Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

Commit:
[PyTorch Edge][type] Add type support for NamedTuple custom class (import)

Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -164,23 +167,190 @@ class TypeParser {
+    ((('name', ''), ('type', 'Tuple[Tensor]'), ('default_value', None)),)))))
+
+*/
+class CustomTypeParser final {
```
Have you run any BSB to check the size regress? Is it possible to put it in another file, so that it's easier for the modularization?

cccclai replied:
Yeah for ios,
Compressed  : +3.8 KiB (+0.01%) => 73.69 MiB
Uncompressed: +8.7 KiB (+0.01%) => 161.83 MiB

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -164,23 +167,190 @@ class TypeParser {
+  void registerParserFunction(
+      const std::string& custom_type,
+      const std::function<TypePtr(IValue&)>& custom_type_parser_function) {
+    parserFunctions()[custom_type] = custom_type_parser_function;
```
If it's registered, it would make more sense to put it into a separate cpp file, so that it can be optionally built into the binary.

cccclai replied:
Not applicable to the new implementation.

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Add a CustomTypeParser to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
('type_name',
    ('NamedTuple,
        ('filed_name_1', 'field_type_1')
        ('filed_name_2', 'field_type_2')
    )
)
```
This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```

This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```

This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```

This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```

This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```

This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -38,6 +39,42 @@ class TypeParser {
     lex();
   }
 
+  explicit TypeParser(std::vector<std::string>& pythonStrs)
```
The same question as the PR below: what's the size regress for those functions?

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -127,8 +171,95 @@ class TypeParser {
+      TypePtr field_type = parse();
+      field_names.emplace_back(field_name);
+      field_types.emplace_back(field_type);
+      std::cout << cur() << std::endl;
```
Remove?

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -127,8 +171,95 @@ class TypeParser {
+      expect("[");
+      std::string field_name = next();
+      expect(",");
+      TypePtr field_type = parse();
```
Nice to support nested types and reuse the existing parsing!

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/type_parser.cpp:
```diff
@@ -37,6 +39,42 @@ TypeParser::TypeParser(std::string pythonStr)
+
+// For the Python string list parsing, the order of the Python string matters.
+// In bytecode, the order of the type list correspondings to the order of
+// instruction. In nested type, the lowest level type will be at the beginning
```
It would be great to order in export side, so that when importing, we don't worry about ordering and just simply parse the ordered types.

cccclai replied:
Currently the export side orders the list correctly. This comment means if the list from export is shuffled, likely it won't be parsed correctly.

iseeyuan commented on test/cpp/jit/test_mobile_type_parser.cpp:
```diff
@@ -41,6 +38,110 @@ TEST(MobileTypeParserTest, NestedContainersAnnotationStrWithSpaces) {
+      "      [teacher_label, Tensor]"
+      "      ]");
+  std::string type_str_2(
+      "Dict[str, __torch__.base_models.preproc_types.PreprocOutputType]");
```
Got it here. Thanks!

cccclai commented on test/cpp/jit/test_mobile_type_parser.cpp:
```diff
@@ -41,6 +38,110 @@ TEST(MobileTypeParserTest, NestedContainersAnnotationStrWithSpaces) {
   ASSERT_EQ(tuple_ps, tuple_space_tps);
 }
 
+TEST(MobileTypeParserTest, NamedTuple) {
```
Yeah it can work. See DictNestedNamedTupleTypeList

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge][type] Add type support for NamedTuple custom class (import)"




Extend `type_parser` to handle `NamedTuple` type. It can be extended to handle other types when needed. The custom type will follow the following format:
```
"qualified_named[
    NamedTuple, [
        [filed_name_1, field_type_1],
        [filed_name_2, field_type_2]
    ]
]"
```
For example:
```
"__torch__.base_models.sparse_nn.pytorch_preproc_types.PreprocOutputType[
    NamedTuple, [
        [float_features, Tensor],
        [id_list_features, List[Tensor]],
        [label,  Tensor],
        [weight, Tensor],
        ]
    ]"
```

For nested types, the order of type lists from type table should be:
```
std::string type_1 = “__torch__.C [
    NamedTuple, [
        [field_name_c_1, Tensor],
        [field_name_c_2, Tuple[Tensor, Tensor]],
    ]
]”

std::string type_2 = “__torch__.B [
   NamedTuple, [
       [field_name_b, __torch__.C ]
   ]
]”

std::string type_3 = “__torch__.A[
   NamedTuple, [
       [field_name_a, __torch__.B]
   ]
]”
std::vector<std::string> type_strs = {type_str_1, type_str_2, type_3};
std::vector<TypePtr> type_ptrs =  c10::parseType(type_strs);
```

namedtuple from both `collection` and `typing` are supported
```

from typing import NamedTuple
from collections import namedtuple
```



This change only adds the parser and now new runtime can read the above format.

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547/)

Differential Revision: [D30261547](https://our.internmc.facebook.com/intern/diff/D30261547)

[ghstack-poisoned]
